### PR TITLE
Added LIS3MDL compass driver

### PR DESCRIPTION
**Pull request by tridge**
This adds a driver for the LIS3MDL magnetometer. 
The PR has so many patches as I built on top of the fast-sampling2 PR

I've thrown in an AK09916 driver as a sweetener

### CODE REVIEW

File: libraries/AP_Compass/AP_Compass.cpp [hunk 1]
```diff
@@ -499,73 +511,73 @@ void Compass::_detect_backends(void)
...
-    _add_backend(AP_Compass_HMC5843::probe(*this, hal.i2c_mgr->get_device(HAL_COMPASS_HMC5843_I2C_BUS, HAL_COMPASS_HMC5843_I2C_ADDR), true),
+    ADD_BACKEND(AP_Compass_HMC5843::probe(*this, hal.i2c_mgr->get_device(HAL_COMPASS_HMC5843_I2C_BUS, HAL_COMPASS_HMC5843_I2C_ADDR), true),
                  AP_Compass_HMC5843::name, true);
-    _add_backend(AP_Compass_LSM303D::probe(*this, hal.spi->get_device("lsm9ds0_am")),
+    ADD_BACKEND(AP_Compass_LSM303D::probe(*this, hal.spi->get_device("lsm9ds0_am")),
                  AP_Compass_LSM303D::name, false);
 #elif HAL_COMPASS_DEFAULT == HAL_COMPASS_BH
     // In BH, only one compass should be detected
-    bool ret = _add_backend(AP_Compass_HMC5843::probe(*this, hal.i2c_mgr->get_device(HAL_COMPASS_HMC5843_I2C_BUS, HAL_COMPASS_HMC5843_I2C_ADDR)),
+    bool ret = ADD_BACKEND(AP_Compass_HMC5843::probe(*this, hal.i2c_mgr->get_device(HAL_COMPASS_HMC5843_I2C_BUS, HAL_COMPASS_HMC5843_I2C_ADDR)),
```
Comment (OXINARF) [inline libraries/AP_Compass/AP_Compass.cpp]: This doesn't work and is the cause for Semaphore to be failing.

**Comment by OXINARF** [comment]
This only adds these two new drivers to PX4 boards, which is probably fine for now, but we should think of a way to improve this and detect all drivers in the external buses in all boards.

File: APMrover2/GCS_Mavlink.cpp [hunk 2]
```diff
@@ -725,7 +741,7 @@ void GCS_MAVLINK_Rover::handleMessage(mavlink_message_t* msg)
         }
         mavlink_statustext_t packet;
         mavlink_msg_statustext_decode(msg, &packet);
-        char text[MAVLINK_MSG_STATUSTEXT_FIELD_TEXT_LEN+1+4] = { 'G','C','S',':'};
+        char text[MAVLINK_MSG_STATUSTEXT_FIELD_TEXT_LEN+1+4] = { 'G', 'C', 'S', ':'};
```
Comment (muramura) [inline APMrover2/GCS_Mavlink.cpp]: I think that it is easier to distinguish clearly from the variable if you insert a space  in +1+4.

**Comment by tridge** [comment]
@OXINARF  I agree. The only issue is how robust the detection is. Some sensors only have an 8 bit WHOAMI register, and some have none at all. It would be easy to get a false positive.

**Comment by tridge** [comment]
applied

**Pull request closed without merging**